apps/governance.mento.org/app/components/lock-list.tsx

```diff
@@ -0,0 +1,619 @@
+    const withdrawnTotal = totalOriginal - totalLockedNow;
+
+    const withdrawnMap = new Map<string, bigint>();
+    let remainingWithdrawn = withdrawnTotal;
+
+    // Initialize all locks with 0 withdrawn amount
+    for (const { lock } of lockData) {
+      withdrawnMap.set(lock.lockId, 0n);
+    }
+
+    // Allocate withdrawn using same logic (expired first, then vesting pro-rata)
+    for (const { lock, amount } of expiredLocks) {
+      if (remainingWithdrawn <= 0n) {
+        withdrawnMap.set(lock.lockId, 0n);
+        continue;
+      }
+      const take = remainingWithdrawn < amount ? remainingWithdrawn : amount;
+      withdrawnMap.set(lock.lockId, take);
+      remainingWithdrawn -= take;
+    }
+
+    if (remainingWithdrawn > 0n && vestingLocks.length > 0) {
+      const totalVested = vestingLocks.reduce(
+        (sum, l) => sum + l.vestedAmount,
+        0n,
+      );
+      if (totalVested > 0n) {
+        let allocatedFromVestingWithdrawn = 0n;
+        for (const { lock, vestedAmount } of vestingLocks) {
+          // Calculate pro-rata allocation for withdrawn amounts
+          const allocation = (vestedAmount * remainingWithdrawn) / totalVested;
+          withdrawnMap.set(
+            lock.lockId,
+            (withdrawnMap.get(lock.lockId) ?? 0n) + allocation,
+          );
+          allocatedFromVestingWithdrawn += allocation;
+        }
+        remainingWithdrawn -= allocatedFromVestingWithdrawn;
+      }
+    }
+
+    // Step 3: Compute remaining amounts
+    const remainingMap = new Map<string, bigint>();
+    for (const { lock, amount } of lockData) {
+      const withdrawn = withdrawnMap.get(lock.lockId) ?? 0n;
+      const remaining = amount > withdrawn ? amount - withdrawn : 0n;
+      remainingMap.set(lock.lockId, remaining);
+    }
+
+    // Ensure remaining total matches chain total
+    const remainingTotal = Array.from(remainingMap.values()).reduce(
+      (a, b) => a + b,
+      0n,
+    );
+    const remainingDust = totalLockedNow - remainingTotal;
+    if (remainingDust !== 0n && lockData.length > 0) {
+      const firstLock = lockData[0];
+      if (firstLock) {
+        const firstLockId = firstLock.lock.lockId;
+        remainingMap.set(
+          firstLockId,
+          (remainingMap.get(firstLockId) ?? 0n) + remainingDust,
+        );
+      }
+    }
+
+    // Return combined estimates with veMENTO calculation
+    const estimates = new Map();
+    for (const { lock, expirySec } of lockData) {
+      const remaining = remainingMap.get(lock.lockId) ?? 0n;
+
+      // Use contract veMENTO value if available, otherwise calculate manually
+      let veMentoAmount = veMentoMap.get(lock.lockId) ?? 0n;
+      // Fallback calculation if contract value not available
+      if (veMentoAmount === 0n && nowSec < expirySec) {
+        const timeRemaining = BigInt(expirySec - nowSec);
+        const totalLockDuration = BigInt(
+          (lock.cliff + lock.slope - 0.25) * WEEK,
```

Comment:
Caveat: the no. of weeks is not 100% precise because the smart contract is counting weeks always from wednesdays. so if you were to create a lock on a tuesday, the first week would technically be just 1 day. if you were to lock on a wednesday, however, the first week is the full 7 days